### PR TITLE
issue 526.5: add analyzer retrieval functionality

### PR DESCRIPTION
**Pull request by vasiliy-arangodb**

### CODE REVIEW

File: arangod/Aql/RemoteExecutor.h [hunk 1]
```diff
@@ -135,6 +135,7 @@ class ExecutionBlockImpl<RemoteExecutor> : public ExecutionBlock {
   /// @brief Mutex to cover against the race, that a getSome request
   ///        is responded before the ticket id is registered.
   arangodb::Mutex _communicationMutex;
+  std::atomic<std::thread::id> _communicationMutexOwner; // current thread owning '_communicationMutex' lock (workaround for non-recusrive MutexLocker)
```
Comment (mchacki) [inline arangod/Aql/RemoteExecutor.h]: OK this is a "fix" i need to reject.

This mutex is used to protect the local state and every place where this mutex is in use cannot be executed (even not using the same thread) while we are in an other place protected by this mutex.

Example:
in getSome we send a message over the ClusterComm.
On return of the ClusterComm (dispatching an asynchrounous call, which allows this code to continue and if the async call is faster it will be held up by exactly this mutex)
Now back the clusterComm will return an OperationID.
This is now retained and we only accept responses for exactly this OperationID.

If now the response is actually injected within the thread and while holding this Mutex (which is the case you run into in your test) the OperationID returned above is NOT yet set.
So the response needs to be rejected (assuming the Mocks are working as intendent) and you cannot get this to work properly.
If the Mocks are not honoring the OperationIDs (which i suspect right now) the mock will break logic depending on this operation IDs in other places.
If we use this RECURSIVE_MUTEX in this code here (even if we just use it in tests) we will NOT see if we get into invalid state, as getting this mutex twice in the thread is disallowed!
We need to see this violation during test phase so we need to keep it a standard mutex.

File: tests/IResearch/IResearchAnalyzerFeature-test.cpp [hunk 2]
```diff
@@ -1155,8 +1146,8 @@ SECTION("test_persistence") {
     arangodb::iresearch::IResearchAnalyzerFeature feature(s.server);
 
     feature.start();
-
-    feature.visit([&expected](irs::string_ref const& name, irs::string_ref const& type, irs::string_ref const& properties)->bool {
+/*
```
Comment (mchacki) [inline tests/IResearch/IResearchAnalyzerFeature-test.cpp]: What is the issue with these MacOS tests?
ArangoDB supports MacOs so ArangoSearch needs to do it too.
If this feature does not work under Mac we cannot include it.
Just removing a test for a production feature is not a fix

File: arangod/Aql/RemoteExecutor.cpp [hunk 3]
```diff
@@ -34,6 +34,63 @@
 #include <lib/Rest/CommonDefines.h>
 #include <velocypack/Iterator.h>
 #include <velocypack/velocypack-aliases.h>
+namespace {
+
+// identical code to RecursiveMutexLocker in ClusterInfo.cpp
```
[the remark this answers is not in the document]
Comment (vasiliy-arangodb) [reply arangod/Aql/RemoteExecutor.cpp]:
```
> As far as i see this mutex locker class is completely independend of any feature from this place
true
> so if at all it should be moved somewhere into lib/
the question is where, when vocbase.cpp and ClusterInfo.cpp were modified there was no answer to this question, hence the code was just left in their respective classes
```

File: arangod/Aql/RemoteExecutor.h (hunk 1 above)
Comment (vasiliy-arangodb) [reply arangod/Aql/RemoteExecutor.h]:
```
> OK this is a "fix" i need to reject.
how do you recommend to test the functionality currently covered by the test?
or should the functionality not have test coverage?
> If now the response is actually injected within the thread and while holding this Mutex (which is the case you run into in your test) the OperationID returned above is NOT yet set.
So the response needs to be rejected (assuming the Mocks are working as intendent) 
I expect the responses generated by the Mock to be returned to the caller
according to the current code of RemoteExecutor there is no reason to reject the responses if their OperationIDs are what is expected
> If the Mocks are not honoring the OperationIDs
the Mocks and the test data are what explicitly set the OperationIDs, there is nothing to not-honour
> If we use this RECURSIVE_MUTEX in this code here (even if we just use it in tests) we will NOT see if we get into invalid state,
that is why I asked you initially if it was ever intended from regular runtime callflow to be possible to call the two methods from the same thread
you told me "if this would happen in production it is broken, so it cannot call it in the same thread"
```

File: tests/IResearch/IResearchAnalyzerFeature-test.cpp (hunk 2 above)
Comment (vasiliy-arangodb) [reply tests/IResearch/IResearchAnalyzerFeature-test.cpp]:
```
> What is the issue with these MacOS tests?
only on MacOS (not win32 on linux) for some reason some other analyzer is also returned during the visitation
I suspect it has to do with the yet-to-be modified code to make this feature request complete, to come in the subsequent pull-requests
```

File: arangod/IResearch/IResearchAnalyzerFeature.cpp [hunk 4]
```diff
@@ -468,28 +473,81 @@ std::pair<irs::string_ref, irs::string_ref> splitAnalyzerName( // split name
 }
 
 ////////////////////////////////////////////////////////////////////////////////
-/// @brief validate that features are supported by arangod an ensure that their
-///        dependencies are met
+/// @brief read analyzers from vocbase
+/// @return visitation compleated fully
```
Comment (jsteemann) [inline arangod/IResearch/IResearchAnalyzerFeature.cpp]:
```suggestion
/// @return visitation completed fully
```

File: arangod/IResearch/IResearchAnalyzerFeature.cpp [hunk 5]
```diff
@@ -468,28 +473,81 @@ std::pair<irs::string_ref, irs::string_ref> splitAnalyzerName( // split name
...
-          TRI_ERROR_BAD_PARAMETER, // code
-          std::string("missing feature '") + std::string(irs::frequency::type().name()) +"' required when '" + std::string(feature->name()) + "' feature is specified"
-        );
-      }
-    } else if (feature) {
+arangodb::Result visitAnalyzers( // visit analyzers
+  TRI_vocbase_t& vocbase, // vocbase to visit
+  std::function<arangodb::Result(arangodb::velocypack::Slice const& slice)> const& visitor // visitor
+) {
+  if (arangodb::ServerState::instance()->isClusterRole()) {
```
Comment (jsteemann) [inline arangod/IResearch/IResearchAnalyzerFeature.cpp]: Should this be `isCoordinator()` instead?
Does the `_analyzers` collection exist on DB servers at all? Its name there is probably `s100xxx` or something.

File: arangod/IResearch/IResearchAnalyzerFeature.cpp [hunk 6]
```diff
@@ -468,28 +473,81 @@ std::pair<irs::string_ref, irs::string_ref> splitAnalyzerName( // split name
...
+    arangodb::AccessMode::Type::READ // access more
+  );
+  auto res = trx.begin();
+
+  if (!res.ok()) {
+    return res;
+  }
+
+  auto commit  = irs::make_finally([&trx]()->void { trx.commit(); }); // end read-only transaction
+  auto result = trx.all(ANALYZER_COLLECTION_NAME, 0, 2, options);
```
Comment (jsteemann) [inline arangod/IResearch/IResearchAnalyzerFeature.cpp]: why are we reading at most 2 documents from the collection here?

File: arangod/IResearch/IResearchAnalyzerFeature.cpp [hunk 7]
```diff
@@ -957,6 +1064,21 @@ IResearchAnalyzerFeature::AnalyzerPool::ptr IResearchAnalyzerFeature::get( // fi
...
 ) const noexcept {
   try {
+    auto split = splitAnalyzerName(name);
+
+    if (!split.first.null()) { // do not trigger load for static-analyzer requests
+      auto res = // load analyzers for database
+        const_cast<IResearchAnalyzerFeature*>(this)->loadAnalyzers(split.first);
+
+      if (!res.ok()) {
+        LOG_TOPIC(,WARN, arangodb::iresearch::TOPIC)
```
Comment (jsteemann) [inline arangod/IResearch/IResearchAnalyzerFeature.cpp]: this looks like as if it will not compile. Log ID is missing here.

File: arangod/IResearch/IResearchAnalyzerFeature.cpp [hunk 8]
```diff
@@ -1143,6 +1265,357 @@ IResearchAnalyzerFeature::AnalyzerPool::ptr IResearchAnalyzerFeature::get( // fi
...
+        return arangodb::Result(); // no collection means nothing to load
+      }
+    }
+
+    Analyzers analyzers;
+    auto visitor = [this, &analyzers, &vocbase]( // visitor
+      arangodb::velocypack::Slice const& slice // analyzer definition
+    )->arangodb::Result {
+      if (!slice.isObject()) {
+        LOG_TOPIC(,ERR, arangodb::iresearch::TOPIC)
```
Comment (jsteemann) [inline arangod/IResearch/IResearchAnalyzerFeature.cpp]: log ID missing here.

File: arangod/IResearch/IResearchAnalyzerFeature.cpp [hunk 9]
```diff
@@ -1143,6 +1265,357 @@ IResearchAnalyzerFeature::AnalyzerPool::ptr IResearchAnalyzerFeature::get( // fi
...
+      irs::flags features;
+      irs::string_ref key;
+      irs::string_ref name;
+      irs::string_ref type;
+      irs::string_ref properties;
+      std::string propertiesBuf;
+
+      if (!slice.hasKey(arangodb::StaticStrings::KeyString) // no such field (required)
+          || !slice.get(arangodb::StaticStrings::KeyString).isString()) {
+        LOG_TOPIC(,ERR, arangodb::iresearch::TOPIC)
```
Comment (jsteemann) [inline arangod/IResearch/IResearchAnalyzerFeature.cpp]: log ID missing here.

File: arangod/IResearch/IResearchAnalyzerFeature.cpp [hunk 10]
```diff
@@ -1143,6 +1265,357 @@ IResearchAnalyzerFeature::AnalyzerPool::ptr IResearchAnalyzerFeature::get( // fi
...
+          << "failed to find a string value for analyzer '" << arangodb::StaticStrings::KeyString << "' while loading analyzer form collection '" << ANALYZER_COLLECTION_NAME << "' in database '" << vocbase->name() + "', skipping it: " << slice.toString();
+
+        return arangodb::Result(); // skip analyzer
+      }
+
+      key = getStringRef(slice.get(arangodb::StaticStrings::KeyString));
+
+      if (!slice.hasKey("name") // no such field (required)
+          || !(slice.get("name").isString() || slice.get("name").isNull())) {
+        LOG_TOPIC(,ERR, arangodb::iresearch::TOPIC)
```
Comment (jsteemann) [inline arangod/IResearch/IResearchAnalyzerFeature.cpp]: log ID missing here.

File: arangod/IResearch/IResearchAnalyzerFeature.cpp [hunk 11]
```diff
@@ -1143,6 +1265,357 @@ IResearchAnalyzerFeature::AnalyzerPool::ptr IResearchAnalyzerFeature::get( // fi
...
+          << "failed to find a string value for analyzer 'name' while loading analyzer form collection '" << ANALYZER_COLLECTION_NAME << "' in database '" << vocbase->name() + "', skipping it: " << slice.toString();
+
+        return arangodb::Result(); // skip analyzer
+      }
+
+      name = getStringRef(slice.get("name"));
+
+      if (!slice.hasKey("type") // no such field (required)
+          || !(slice.get("type").isString() || slice.get("name").isNull())) {
```
Comment (jsteemann) [inline arangod/IResearch/IResearchAnalyzerFeature.cpp]: should the second check check attribute "name" or "type"?

File: arangod/IResearch/IResearchAnalyzerFeature.cpp [hunk 12]
```diff
@@ -1143,6 +1265,357 @@ IResearchAnalyzerFeature::AnalyzerPool::ptr IResearchAnalyzerFeature::get( // fi
...
+      if (itr->second // valid new entry
+          && !equalAnalyzer(*(entry.second), itr->second->type(), itr->second->properties(), itr->second->features())) {
+        return arangodb::Result( // result
+          TRI_ERROR_BAD_PARAMETER, // code
+          std::string("name collision detected while registering a duplicate arangosearch analizer name '") + std::string(itr->second->name()) + "' type '" + std::string(itr->second->type()) + "' properties '" + std::string(itr->second->properties()) + "', previous registration type '" + std::string(entry.second->type()) + "' properties '" + std::string(entry.second->properties()) + "'"
+        );
+      }
+
+      itr->second = entry.second; // reuse old analyzer pool to avoid duplicates in memmory
+      const_cast<Analyzers::key_type&>(itr->first) = entry.first; // point key at old pool
```
Comment (jsteemann) [inline arangod/IResearch/IResearchAnalyzerFeature.cpp]: mmmhhh. Why not erase and re-insert with the new key?

File: arangod/IResearch/IResearchAnalyzerFeature.cpp [hunk 13]
```diff
@@ -1933,14 +2311,33 @@ arangodb::Result IResearchAnalyzerFeature::storeAnalyzer(AnalyzerPool& pool) {
...
-  ReadMutex mutex(_mutex);
-  SCOPED_LOCK(mutex);
+) const {
+  if (vocbase) { // do not trigger load for all-databases requests
+    auto res = const_cast<IResearchAnalyzerFeature*>(this)->loadAnalyzers( // load analyzers for database
+      vocbase->name() // args
+    );
+
+    if (!res.ok()) {
+      LOG_TOPIC(,WARN, arangodb::iresearch::TOPIC)
```
Comment (jsteemann) [inline arangod/IResearch/IResearchAnalyzerFeature.cpp]: log ID missing

File: arangod/IResearch/IResearchLinkMeta.cpp [hunk 14]
```diff
@@ -234,7 +234,7 @@ bool IResearchLinkMeta::init( // initialize meta
           auto analyzer = analyzers->get(name);
 
           if (!analyzer) {
-            errorField = fieldName + "=>" + std::string(name);
+            errorField = fieldName + "=>" + value.copyString(); // original (non-normalized) 'name' valie
```
Comment (jsteemann) [inline arangod/IResearch/IResearchLinkMeta.cpp]:
```suggestion
            errorField = fieldName + "=>" + value.copyString(); // original (non-normalized) 'name' value
```

File: tests/Cluster/ClusterInfo-test.cpp [hunk 15]
```diff
@@ -141,7 +141,8 @@ struct ClusterInfoSetup {
     arangodb::EngineSelectorFeature::ENGINE = &engine;
 
     // suppress INFO {authentication} Authentication is turned on (system only), authentication for unix sockets is turned on
-    arangodb::LogTopic::setLogLevel(arangodb::Logger::AUTHENTICATION.name(), arangodb::LogLevel::WARN);
+    // suppress WARNING {authentication} --server.jwt-secret is insecure. Use --server.jwt-secret-keyfile instead
+    arangodb::LogTopic::setLogLevel(arangodb::Logger::AUTHENTICATION.name(), arangodb::LogLevel::ERR);
```
Comment (jsteemann) [inline tests/Cluster/ClusterInfo-test.cpp]: The superfluous error message issue should have been fixed by now.
Thus can we revert the change to the log level here?

File: tests/IResearch/ClusterCommMock.cpp [hunk 16]
```diff
@@ -61,29 +61,37 @@ arangodb::OperationID ClusterCommMock::asyncRequest(
...
    arangodb::ClusterCommTimeout initTimeout
 ) {
-  auto entry = _requests.emplace(
-    std::piecewise_construct,
-    std::forward_as_tuple(++_operationId), // non-zero value
-    std::forward_as_tuple(coordTransactionID, destination, reqtype, path, body, headerFields, callback, singleRequest)
-  );
+  // execute before insertion to avoid consuming an '_operationId'
+  if (arangodb::rest::RequestType::PUT == reqtype
+      && std::string::npos != path.find("/_api/aql/shutdown/")) {
```
Comment (jsteemann) [inline tests/IResearch/ClusterCommMock.cpp]: Do this affect any of the AQL catch tests?

File: tests/IResearch/ExecutionBlockMock-test.cpp [hunk 17]
```diff
@@ -87,8 +87,8 @@ struct IResearchBlockMockSetup {
     arangodb::tests::init(true);
 
     // suppress INFO {authentication} Authentication is turned on (system only), authentication for unix sockets is turned on
-    arangodb::LogTopic::setLogLevel(arangodb::Logger::AUTHENTICATION.name(),
-                                    arangodb::LogLevel::WARN);
+    // suppress WARNING {authentication} --server.jwt-secret is insecure. Use --server.jwt-secret-keyfile instead
+    arangodb::LogTopic::setLogLevel(arangodb::Logger::AUTHENTICATION.name(), arangodb::LogLevel::ERR);
```
Comment (jsteemann) [inline tests/IResearch/ExecutionBlockMock-test.cpp]: can this change be reverted now? I think the original error with the superfluous log messages is fixed already.

File: tests/IResearch/ExpressionFilter-test.cpp [hunk 18]
```diff
@@ -265,7 +265,8 @@ struct IResearchExpressionFilterSetup {
     arangodb::tests::init(true);
 
     // suppress INFO {authentication} Authentication is turned on (system only), authentication for unix sockets is turned on
-    arangodb::LogTopic::setLogLevel(arangodb::Logger::AUTHENTICATION.name(), arangodb::LogLevel::WARN);
+    // suppress WARNING {authentication} --server.jwt-secret is insecure. Use --server.jwt-secret-keyfile instead
+    arangodb::LogTopic::setLogLevel(arangodb::Logger::AUTHENTICATION.name(), arangodb::LogLevel::ERR);
```
Comment (jsteemann) [inline tests/IResearch/ExpressionFilter-test.cpp]: same

File: tests/IResearch/IResearchDocument-test.cpp [hunk 19]
```diff
@@ -138,7 +138,8 @@ struct IResearchDocumentSetup {
     arangodb::tests::init();
 
     // suppress INFO {authentication} Authentication is turned on (system only), authentication for unix sockets is turned on
-    arangodb::LogTopic::setLogLevel(arangodb::Logger::AUTHENTICATION.name(), arangodb::LogLevel::WARN);
+    // suppress WARNING {authentication} --server.jwt-secret is insecure. Use --server.jwt-secret-keyfile instead
+    arangodb::LogTopic::setLogLevel(arangodb::Logger::AUTHENTICATION.name(), arangodb::LogLevel::ERR);
```
Comment (jsteemann) [inline tests/IResearch/IResearchDocument-test.cpp]: same (and also for all following instances of this change)

File: arangod/IResearch/IResearchAnalyzerFeature.cpp (hunk 5 above)
Comment (vasiliy-arangodb) [reply arangod/IResearch/IResearchAnalyzerFeature.cpp]: this should be coordinator or db-server, as both of them should gather their data from the entire cluster

File: arangod/IResearch/IResearchAnalyzerFeature.cpp (hunk 7 above)
Comment (vasiliy-arangodb) [reply arangod/IResearch/IResearchAnalyzerFeature.cpp]: compiles and runs perfectly

File: arangod/IResearch/IResearchAnalyzerFeature.cpp (hunk 8 above)
Comment (vasiliy-arangodb) [reply arangod/IResearch/IResearchAnalyzerFeature.cpp]: it's empty

File: tests/Cluster/ClusterInfo-test.cpp (hunk 15 above)
Comment (vasiliy-arangodb) [reply tests/Cluster/ClusterInfo-test.cpp]: it's safe to ignore all errors from the authentication topic anyway, since there are tests to check invalid permissions
and there are quite a few test files with this change, hence I am not sure if it's worthwhile to revert

File: tests/IResearch/ClusterCommMock.cpp (hunk 16 above)
Comment (vasiliy-arangodb) [reply tests/IResearch/ClusterCommMock.cpp]: it is required for the new CATCH tests that were added since the 'shutdown' requests ended up consuming all the test responses (that were meant to match subsequent requests) hence the mock code forcefully terminates such requests

File: tests/IResearch/ExecutionBlockMock-test.cpp (hunk 17 above)
Comment (vasiliy-arangodb) [reply tests/IResearch/ExecutionBlockMock-test.cpp]: it's safe to ignore all errors from the authentication topic anyway, since there are tests to check invalid permissions
and there are quite a few test files with this change, hence I am not sure if it's worthwhile to revert

File: tests/IResearch/ExpressionFilter-test.cpp (hunk 18 above)
Comment (vasiliy-arangodb) [reply tests/IResearch/ExpressionFilter-test.cpp]: it's safe to ignore all errors from the authentication topic anyway, since there are tests to check invalid permissions
and there are quite a few test files with this change, hence I am not sure if it's worthwhile to revert

File: tests/IResearch/IResearchDocument-test.cpp (hunk 19 above)
Comment (vasiliy-arangodb) [reply tests/IResearch/IResearchDocument-test.cpp]: it's safe to ignore all errors from the authentication topic anyway, since there are tests to check invalid permissions
and there are quite a few test files with this change, hence I am not sure if it's worthwhile to revert

File: arangod/IResearch/IResearchAnalyzerFeature.cpp (hunk 6 above)
Comment (vasiliy-arangodb) [reply arangod/IResearch/IResearchAnalyzerFeature.cpp]: looks like a typo, but the code ignores that argument anyway, hence all the tests passed
changed to '0'

File: arangod/IResearch/IResearchAnalyzerFeature.cpp (hunk 4 above)
Comment (vasiliy-arangodb) [reply arangod/IResearch/IResearchAnalyzerFeature.cpp]: Done

**Comment by vasiliy-arangodb** [comment]
http://jenkins01.arangodb.biz:8080/job/arangodb-matrix-pr/3822/

File: arangod/IResearch/IResearchAnalyzerFeature.cpp (hunk 7 above)
Comment (jsteemann) [reply arangod/IResearch/IResearchAnalyzerFeature.cpp]: However, this is not how the LOG_TOPIC macro is supposed to be used.

File: arangod/IResearch/IResearchAnalyzerFeature.cpp (hunk 8 above)
Comment (jsteemann) [reply arangod/IResearch/IResearchAnalyzerFeature.cpp]: see above. This is not how it is supposed to be used.

File: tests/Cluster/ClusterInfo-test.cpp (hunk 15 above)
Comment (jsteemann) [reply tests/Cluster/ClusterInfo-test.cpp]: Ok, fine. It's a change unrelated to the subject of this PR, but it doesn't do any harm here.

**Comment by jsteemann** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/3830/
